pkg/ansible/generate.go

```diff
@@ -140,79 +144,243 @@ openshift_aws_create_s3: false
+      cidr_ip: 0.0.0.0/0
+    - proto: all
+      from_port: all
+      to_port: all
+      group_name: "{{ openshift_aws_clusterid }}"
+  [[ .Name ]]:
+    name: "{{ openshift_aws_clusterid }}_compute"
+    desc: "{{ openshift_aws_clusterid }} compute node instances"
+`
+	DefaultInventory = `
+[OSEv3:children]
+masters
+nodes
+etcd
+
+[OSEv3:vars]
+ansible_become=true
+
+[masters]
+
+[etcd]
+
+[nodes]
+`
 )
 
-type VarsGenerator struct {
-	Vars map[string]string
+type clusterParams struct {
+	Name       string
+	Region     string
+	SSHKeyName string
+	DefaultAMI string
 }
 
-func NewVarsGenerator(cluster *coapi.Cluster) *VarsGenerator {
-
-	// Any vars that change from cluster to cluster need to be defined here:
-	inv := &VarsGenerator{
-		Vars: map[string]string{
-			cloudProviderKindVar: "aws",
-			awsRegionVar:         "us-east-1",
-			awsClusterIDVar:      cluster.Name,
-			elbBasenameVar:       cluster.Name,
-			awsVPCNameVar:        cluster.Name,
-			awsAMIVar:            fmt.Sprintf("%s-ami-base", cluster.Name),
-		},
-	}
-	return inv
+type machineSetParams struct {
+	Name         string
+	Size         int
+	AMIName      string
+	InstanceType string
 }
 
-// GenerateVars returns a string suitable for use as a vars file when calling openshift-ansible pods.
-func (ig *VarsGenerator) GenerateVars() (string, error) {
+func GenerateClusterVars(cluster *coapi.Cluster) (string, error) {
+
+	// Currently only AWS is supported. If we don't have an AWS cluster spec, return an error
+	if cluster.Spec.Hardware.AWS == nil {
+		return "", fmt.Errorf("no AWS spec found in the cluster, only AWS is currently supported")
+	}
+
 	// Change template delimiters to avoid conflict with {{ }} use in ansible vars:
-	t, err := template.New("varsfilecontents").Delims("[[", "]]").Parse(varsTemplate)
+	t, err := template.New("clustervars").Delims("[[", "]]").Parse(clusterVarsTemplate)
 	if err != nil {
 		return "", err
 	}
 
+	params := clusterParams{
+		Name:       cluster.Name,
+		Region:     cluster.Spec.Hardware.AWS.Region,
+		SSHKeyName: cluster.Spec.Hardware.AWS.KeyPairName,
+	}
+
+	if cluster.Spec.DefaultHardwareSpec != nil && cluster.Spec.DefaultHardwareSpec.AWS != nil {
+		params.DefaultAMI = cluster.Spec.DefaultHardwareSpec.AWS.AMIName
+	}
+
 	var buf bytes.Buffer
-	err = t.Execute(&buf, ig.Vars)
+	err = t.Execute(&buf, params)
 	if err != nil {
 		return "", err
 	}
 	return buf.String(), nil
 }
+
+func GenerateMachineSetVars(cluster *coapi.Cluster, machineSet *coapi.MachineSet) (string, error) {
+	commonVars, err := GenerateClusterVars(cluster)
+	if err != nil {
+		return "", err
+	}
+
+	if machineSet.Spec.Hardware.AWS == nil {
+		return "", fmt.Errorf("no AWS spec found on machine set, only AWS is currently supported")
+	}
+
+	var buf bytes.Buffer
+	buf.WriteString(commonVars)
+
+	var varsTemplate string
+	switch {
+	case machineSet.Spec.NodeType == coapi.NodeTypeMaster:
+		varsTemplate = masterVarsTemplate
+	case machineSet.Spec.Infra:
```

Comment:
How do we handle the case where the machineset is both master and infra? Overall, I am a little confused about how each of the vars templates work. I would have expected (1) that there some base vars that each machineset type would include, (2) that some vars would be added for master machinesets, (3) that some vars would be added for non-master machinesets, and (4) that some vars would be added for infra machinesets.

Reply:
It is currently not possible with the playbooks we have to have the master be infra. Playbooks are designed to work with 3 distinct sets of nodes (based on my conversation with @abutcher)

Reply:
Do we expect that to change in the future? Or do we need to adjust the API types to reflect that reality?

Reply:
I needed to duplicate some structs from the role defaults so that I could have a separate set of structs per machineset. As it is now, you're expected to create all non-master asg's at the same time.

Reply:
@staebler we need to have the conversation and make a call. It comes down to whether we care enough about the use case to justify the work required  to make it work.

Reply:
I'm thinking we should do what we can to at least keep that door open. Spinning up single master all-in-one clusters  could we very desirable for kick the tires scenarios.